pipeline/execout/cachev1/cache.go

```diff
@@ -0,0 +1,329 @@
+}
+
+func (c *OutputCache) String() string {
+	return c.Store.ObjectURL("")
+}
+
+func (c *OutputCache) ListContinuousCacheRanges(ctx context.Context, from uint64) (block.Ranges, error) {
+	cachedRanges, err := c.ListCacheRanges(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("listing cached ranges %q: %w", c.ModuleName, err)
+	}
+	out := listContinuousCacheRanges(cachedRanges, from)
+	return out, nil
+}
+
+func (c *OutputCache) ListCacheRanges(ctx context.Context) (block.Ranges, error) {
+	var out block.Ranges
+	err := derr.RetryContext(ctx, 3, func(ctx context.Context) error {
+		if err := c.Store.Walk(ctx, "", func(filename string) (err error) {
+			r, err := fileNameToRange(filename)
+			if err != nil {
+				return fmt.Errorf("getting range from filename: %w", err)
+			}
+			out = append(out, r)
+			return nil
+		}); err != nil {
+			return fmt.Errorf("walking cache ouputs: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	sort.Slice(out, func(i, j int) bool {
+		return out[i].StartBlock < out[j].StartBlock
+	})
+
+	return out, nil
+}
+
+func (c *OutputCache) Delete(blockID string) {
+	c.Lock()
+	defer c.Unlock()
+
+	delete(c.kv, blockID)
+}
+
+func listContinuousCacheRanges(cachedRanges block.Ranges, from uint64) block.Ranges {
+	cachedRangeCount := len(cachedRanges)
+	var out block.Ranges
+	for i, r := range cachedRanges {
+		if r.StartBlock < from {
+			continue
+		}
+		out = append(out, r)
+		if cachedRangeCount > i+1 {
+			next := cachedRanges[i+1]
+			if next.StartBlock != r.ExclusiveEndBlock { //continuous seq broken
+				break
+			}
+		}
+	}
+
+	return out
+}
+
+func fileNameToRange(filename string) (*block.Range, error) {
+	res := cacheFilenameRegex.FindAllStringSubmatch(filename, 1)
+	if len(res) != 1 {
+		return nil, fmt.Errorf("invalid output cache filename, %q", filename)
+	}
+
+	start := uint64(utils.MustAtoi(res[0][1]))
+	end := uint64(utils.MustAtoi(res[0][2]))
+
+	return &block.Range{
+		StartBlock:        start,
+		ExclusiveEndBlock: end,
+	}, nil
+}
+
+func findBlockRange(ctx context.Context, store dstore.Store, prefixStartBlock uint64) (*block.Range, bool, error) {
+	var exclusiveEndBlock uint64
+
+	paddedBlock := pad(prefixStartBlock)
+
+	var files []string
+	err := derr.RetryContext(ctx, 3, func(ctx context.Context) (err error) {
+		files, err = store.ListFiles(ctx, paddedBlock, math.MaxInt64)
+		return
+	})
+	if err != nil {
+		return nil, false, fmt.Errorf("walking prefix for padded block %s: %w", paddedBlock, err)
+	}
+
+	if len(files) == 0 {
+		return nil, false, nil
+	}
+
+	biggestEndBlock := uint64(0)
+
+	for _, file := range files {
+		endBlock, err := getExclusiveEndBlock(file)
+		if err != nil {
+			return nil, false, fmt.Errorf("getting exclusive end block from file %s: %w", file, err)
+		}
+		if endBlock > biggestEndBlock {
+			biggestEndBlock = endBlock
+		}
+	}
+
+	exclusiveEndBlock = biggestEndBlock
+
+	return block.NewRange(prefixStartBlock, exclusiveEndBlock), true, nil
+}
```

Comment:
I like these functions here. They abstract complexities away, hide them behind a meaningful name. And simplify the call site, all the while not introducing new objects with side effects mutating struct members.

Functions, functions functions! Let's do them more :)